src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,299 @@
+ * algorithm is an iterative algorithm, where in each iteration a single
+ * sampled point is used to perform an update as done in Stochastic Gradient
+ * Descent. For PCA this means iteratively using vectors sampled from a
+ * distribution to update the subspace.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @inproceedings{Arora2012,
+ *   author    = {Arora, Raman and Cotter, Andrew and Livescu, Karen
+ *                and Srebro, Nathan},
+ *   title     = {Stochastic optimization for PCA and PLS.},
+ *   booktitle = {Allerton Conference},
+ *   pages     = {861-868},
+ *   year      = {2012},
+ * }
+ * @endcode
+ */
+template<typename OptimizerType>
+class StochasticPolicyType
+{
+ public:
+  /**
+   * Use the stochastic gradient method to perform the principal components
+   * analysis (PCA).
+   *
+   * @param optimizer Instantiated optimizer used to update the subspace.
+   */
+  StochasticPolicyType(const OptimizerType& optimizer = OptimizerType()) :
+      optimizer(optimizer)
+  {
+    /* Nothing to do here. */
+  }
+
+  /**
+   * Apply Principal Component Analysis to the provided data set using the
+   * exact SVD method.
+   *
+   * @param data Data matrix.
+   * @param centeredData Centered data matrix.
+   * @param transformedData Matrix to put results of PCA into.
+   * @param eigVal Vector to put eigenvalues into.
+   * @param eigvec Matrix to put eigenvectors (loadings) into.
+   * @param rank Rank of the decomposition.
+   */
+  void Apply(const arma::mat& data,
+             const arma::mat& centeredData,
+             arma::mat& transformedData,
+             arma::vec& eigVal,
+             arma::mat& eigvec,
+             const size_t rank)
+  {
+    PCAFunction f(centeredData);
+
+    // Adjust the step size if necessary.
+    SetStepSize(optimizer);
+
+    eigvec = arma::randn(centeredData.n_rows, rank);
+    optimizer.Optimize(f, eigvec);
+
+    // Project the samples to the principals.
+    transformedData = arma::trans(eigvec) * centeredData;
+
+    // Approximate eigenvalues and eigenvectors using Rayleigh–Ritz method.
+    arma::mat u, v;
+    arma::svd_econ(u, eigVal, v, transformedData);
+
+    // Now we must square the singular values to get the eigenvalues.
+    // In addition we must divide by the number of points, because the
+    // covariance matrix is X * X' / (N - 1).
+    eigVal %= eigVal / (data.n_cols - 1);
+  }
+
+  //! Get the optimizer.
+  OptimizerType Optimizer() const { return optimizer; }
+  //! Modify the optimizer.
+  OptimizerType& Optimizer() { return optimizer; }
+
+ private:
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& optimizer)
+  {
+    if (optimizer.StepSize() > 0)
+    {
+      Log::Warn << "StepSize(): invalid value (> 0), automatically take the "
+          << "negative direction." << std::endl;
+      optimizer.StepSize() *= -1;
```

Comment:
Hmm, would it be easier to simply negate the objective function?  The formulation of stochastic PCA here is a maximization, but the mlpack optimizers minimize.  So if you just use the negative objective (and derive the gradient accordingly) I think that removes the need for this.